accepted/2020/reproducible-builds.md

```diff
@@ -0,0 +1,238 @@
+- Latest Roslyn compiler which stores compiler flags into the PDB
+- PDB is automatically included in the NuGet package
+- Strong name keys are checked into source control
+
+After building and publishing the package using the latest SDK, NuGet.org validates the library and assigns a higher score.
+
+#### Developer using NuGet.org
+
+A developer is searching for a logging library on NuGet.org. They search and see two interesting libraries, one with a "verified" badge, and the other without. The developer picks the verified one, and feels confident that the package they're using does not contain code they are not aware of and have visibility to. They trust NuGet to have validated that the package meets the specified supply chain requirements.
+
+#### Corporate / Government developer at a large organization
+
+A developer in an internal team building a service needs to make sure they comply with their supply chain requirements. They use a feed containing only validated packages provided by their internal InfoSec team.
+
+#### NuGet
+
+NuGet is looking at ways to improve the security of the ecosystem and provide metrics around project/package health. Ensuring that the binaries in the project match the source is an important part, ensuring no tampering occurs in the build pipeline.
+
+NuGet will add a validation step on package ingestion:
+
+1. For each managed binary in the NuGet package, call a tool to check the status. Result is yes/no.
+2. If all of the files are successful, the package is marked as reproducible. If a package contains unmanaged binaries, then the result would be interterminite and marked accordingly.
+3. For each managed binary or PDB file in the package, add it to the NuGet symbol server. NuGet will retroactively do this step for all existing packages to-date.
+
+NuGet will allow packages that are not reproducible to be ingested; failure is not a hard block. NuGet will display the reproducible status in appropriate metadata and UI so users and tools that trust NuGet can make informed decisions without revalidating themselves.
+
+## Requirements
+
+### Goals
+
+**P0**
+
+- The validation tool must work on any managed binary, regardless of target framework
+- The validation tool must work on any operating system.
+- The validation tool must support public sources for sources and symbols.
+- The validation tool must be easy to use: single command the user invokes.
+
+**P1**
+
+- The validation tool should support non-public source and symbol locations given credentials.
+
+**P2**
+
+- The validation tool should support F#
+- The validation tool should support any .NET SDK tool which creates or modifies DLL / EXEs
+
+**Future**
+
+- Support for IL weavers
+
+### Non-Goals
+
+- Evaluating provenance of embedded resources, such as .resx or .baml.
+- Support for modification or patching during validation, to create a runnable binary that's functionally different from the original.
+- Support for exporting a rebuilt artifact
```

Comment:
Doesn't this seem counter to this part in the Overviews? 

> Also, should package consumers find a zero-day exploit, they usually have no way to patch the binaries themselves. They must find the corresponding repo, get the necessary dependencies for building it, somehow find the correct version of the code, patch it, and then rebuild. Not an easy task under ideal circumstances, much less viable when time is of the essence.

If the tool doesn't produce (and reproduce) the original environment and the rebuilt artifact, this part of the user story is still not addressed.

Reply:
The explicit goal for v1 is to enable provenance traceability between the binary and the source. In order to do that, the dll will be rebuilt and the result compared, however using the toolchain for patching isn't an explicit goal initially.

Other issues like source generators arise where the source generator won't be re-run here. Depending on what the source generator does, making a change to the source may or may not work.